### PR TITLE
Remove __LinkCustomCRT and crt.cpp

### PR DESCRIPTION
**Pull request by kasper3**
This was the only C++ file under Windows.

### CODE REVIEW

**Comment by stephentoub** [comment]
What's the motivation for this change?

**Comment by kasper3** [comment]
@stephentoub, in places of .NET Core where native code is used, some files are having .cpp extensions with 90-100% of C code in them. That adds additional dependency on C++ library as in the linker links libstdc++ to the produced binary, which is unnecessary and can be trimmed out to make some parts of the product light-weight. This is part of that general sweep and philosophy here is hugely inspired by @jkotas' idea for CoreRT https://github.com/dotnet/corert/issues/3564, which suggests that there is a potential of even walking an extra mile to achieve non-C++'ism.

As we can see that in this particular case, a file with "CRT" code is having ".cpp" extension and all the code is actually a valid ISO C code. Plus under Windows/clrcompression directory, this is the only C++ file (among 20-30 c files in zlib/zlib-intel subdirectories).

**Comment by jkotas** [comment]
This `crt.cpp` file is only used by a special clrcompression build for uap/uapaot flavor (this flavor is used by .NET Native for UWP apps). It existed to handle issues with CRT availability in the early days of the .NET Native project. It should not be needed anymore. I think that the right fix is to delete this cruft and just link against regular CRT.

https://github.com/dotnet/corefx/pull/26229#discussion_r161114326 has some discussion related to this.

In either case, the challenge is testing it without access to internal Microsoft network since .NET Native is not fully opened sourced. Perhaps you can create issue about this, prepare PR with the change, do as much testing as you can think of yourself e.g. by replacing clrcompression.dll in a test UWP app with the one your have built, and then ask for Microsoft person to validate that there are no issues in the closed source part of .NET Native.

cc @ianhays

**Comment by ianhays** [comment]
@jkotas is correct. I was actually about to tackle this myself, so thanks for taking a look @kasper3  :)

All of the __LinkCustomCRT code is now dead and can be removed. The entire crt.cpp file can also be removed.

**Comment by kasper3** [comment]
@ianhays, i have removed it but for some reason, x86 build is failing to link CRT.

File: src/Native/Windows/CMakeLists.txt [hunk 1]
```diff
@@ -91,14 +91,12 @@ endif()
...
-endif()
-
-if (__LinkCustomCRT STREQUAL "true" AND $ENV{__BuildArch} STREQUAL "arm" )
-    set(__LinkArgs "${__LinkArgs} /DEFAULTLIB:ole32.lib")
-elseif(__LinkCustomCRT STREQUAL "true" AND NOT $ENV{__BuildArch} STREQUAL "arm" )
-    set(__LinkArgs "${__LinkArgs} /NODEFAULTLIB")
+    if ($ENV{__BuildArch} STREQUAL "arm" )
+        set(__LinkArgs "${__LinkArgs} /DEFAULTLIB:ole32.lib")
+    else()
+        set(__LinkArgs "${__LinkArgs} /NODEFAULTLIB")
```
Comment (jkotas) [inline src/Native/Windows/CMakeLists.txt]: This `/NODEFAULTLIB` is likely the reason why it does not link. Try removing this line.

File: src/Native/Windows/CMakeLists.txt [hunk 2]
```diff
@@ -113,10 +111,6 @@ set(__LinkArgs "${__LinkArgs} /PDBCOMPRESS")            #shrink pdb size
 set(__LinkArgs "${__LinkArgs} /DEBUG")
 set(__LinkArgs "${__LinkArgs} /IGNORE:4197,4013,4254,4070,4221")
 
-if ($ENV{__BuildArch} STREQUAL "x86" OR $ENV{__BuildArch} STREQUAL "x86_64" )
```
Comment (jkotas) [inline src/Native/Windows/CMakeLists.txt]: What's the reason for this change?

Comment (kasper3) [reply src/Native/Windows/CMakeLists.txt]: Every x86 build in CoreFX is getting this warning:

```
LINK : warning LNK4010: invalid subsystem version number 6.00; default subsystem version assumed
```

then i read https://github.com/dotnet/coreclr/pull/8211#discussion_r88793983 and thought it is not needed.

Comment (AraHaan) [reply src/Native/Windows/CMakeLists.txt]: Oh boy, I wished I knew where it is implemented in C# where it would generate docs for internal / private functions with public ones as well. I thought including doc summaries for internal / private functions would be excloded from generated documentation xml's and used only by Core (in the case of my project) developers. Which would be a great change as well.

Comment (jkotas) [reply src/Native/Windows/CMakeLists.txt]: ok

File: src/Native/Windows/CMakeLists.txt [hunk 3]
```diff
@@ -91,14 +91,10 @@ endif()
...
-    set(__LinkCustomCRT true)
     set(__LinkArgs "${__LinkArgs} /LTCG")
-endif()
-
-if (__LinkCustomCRT STREQUAL "true" AND $ENV{__BuildArch} STREQUAL "arm" )
-    set(__LinkArgs "${__LinkArgs} /DEFAULTLIB:ole32.lib")
-elseif(__LinkCustomCRT STREQUAL "true" AND NOT $ENV{__BuildArch} STREQUAL "arm" )
-    set(__LinkArgs "${__LinkArgs} /NODEFAULTLIB")
+    if ($ENV{__BuildArch} STREQUAL "arm" )
+        set(__LinkArgs "${__LinkArgs} /DEFAULTLIB:ole32.lib")
```
Comment (jkotas) [inline src/Native/Windows/CMakeLists.txt]: Is the explicit `ole32.lib` really needed for arm? It looks odd.

File: src/Native/Windows/CMakeLists.txt [hunk 4]
```diff
@@ -91,16 +91,9 @@ endif()
 
 # App-Container builds require a more specific linking to ensure the APIs imported pass WACK
 if ($ENV{__appContainer} STREQUAL true AND UPPERCASE_CMAKE_BUILD_TYPE STREQUAL "RELEASE")
-    set(__LinkCustomCRT true)
     set(__LinkArgs "${__LinkArgs} /LTCG")
```
Comment (jkotas) [inline src/Native/Windows/CMakeLists.txt]: `/LTCG` option is added a few lines below for all build flavors: https://github.com/dotnet/corefx/pull/30256/files#diff-5822bf65a2871321bce04bde060ce1a0L122

We should not need to duplicate it here.

File: src/Native/Windows/CMakeLists.txt (hunk 1 above)
Comment (ianhays) [reply src/Native/Windows/CMakeLists.txt]: +1. That's the issue. the crt.cpp file is providing the same thing as the default libs, so you need one or the other.

File: src/Native/Windows/CMakeLists.txt (hunk 4 above)
Comment (kasper3) [reply src/Native/Windows/CMakeLists.txt]: Removed

**Comment by ianhays** [comment]
Thanks @kasper3 , looks good!
